File: 2025/09 September/23 Queue Reversal/main.py
# Link: https://www.geeksforgeeks.org/problems/queue-reversal/1


class Solution:
    def reverseQueue(self, q):
        st = []

        while q:
            st.append(q[0])
            q.popleft()

        while st:
            q.append(st.pop())


if __name__ == "__main__":
    from collections import deque

    sol = Solution()
    arr = [5, 10, 15, 20, 25]
    q = deque(arr)
    sol.reverseQueue(q)
    print(f"Reversed Queue: {list(q)}")

    arr = [1, 2, 3, 4, 5]
    q = deque(arr)
    sol.reverseQueue(q)
    print(f"Reversed Queue: {list(q)}")

File: 2025/09 September/23 Queue Reversal/main.cpp
// Link: https://www.geeksforgeeks.org/problems/queue-reversal/1

#include <iostream>
#include <queue>
#include <stack>
using namespace std;

class Solution
{
public:
    void reverseQueue(queue<int> &q)
    {
        stack<int> st;

        while (!q.empty())
        {
            st.push(q.front());
            q.pop();
        }

        while (!st.empty())
        {
            q.push(st.top());
            st.pop();
        }
    }
};

int main()
{
    Solution sol;
    queue<int> q;
    vector<int> elements = {5, 10, 15, 20, 25};
    for (int elem : elements)
    {
        q.push(elem);
    }
    sol.reverseQueue(q);
    cout << "Reversed Queue: ";
    while (!q.empty())
    {
        cout << q.front() << " ";
        q.pop();
    }
    cout << endl;

    elements = {1, 2, 3, 4, 5};
    for (int elem : elements)
    {
        q.push(elem);
    }
    sol.reverseQueue(q);
    cout << "Reversed Queue: ";
    while (!q.empty())
    {
        cout << q.front() << " ";
        q.pop();
    }
    cout << endl;
    return 0;
}